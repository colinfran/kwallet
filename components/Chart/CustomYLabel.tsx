import React, { useEffect } from "react"
import { useSharedValue } from "react-native-reanimated"
import { ChartYLabel } from "@colinfran/animated-charts"

const CustomYLabel = ({
  currentPrice = 0,
  currencySymbol = "$",
  color,
  style,
}): JSX.Element => {
  const symbol = useSharedValue(currencySymbol)
  const latestCurrentPrice = useSharedValue(currentPrice)

  useEffect(() => {
    latestCurrentPrice.value = currentPrice
    symbol.value = currencySymbol
  }, [currencySymbol, currentPrice, latestCurrentPrice, symbol])

  const formatPrice = (value): any => {
    "worklet"
    if (value === "") {
      return `${symbol.value}${Number(latestCurrentPrice.value).toFixed(7)}`
    }
    return `${symbol.value}${Number(value).toFixed(7)}`
  }

  return <ChartYLabel color={color} format={formatPrice} style={style} />
}
export default CustomYLabel
