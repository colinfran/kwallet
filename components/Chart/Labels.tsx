import React, { useCallback, useContext, useMemo, useState } from "react"
import { View, Text } from "react-native"
import { useChartData } from "@colinfran/animated-charts"
import { DataContext } from "../../providers/DataProvider"
import { getCurrencySymbol } from "../../constants/currencies"

const trim = (val: number): number => {
  return Math.min(Math.max(val, 0.05), 0.95)
}

const CenteredLabel = ({ position, style, width, ...props }): JSX.Element => {
  const [componentWidth, setWidth] = useState(0)
  const onLayout = useCallback(
    ({
      nativeEvent: {
        layout: { width: newWidth },
      },
    }) => {
      setWidth(newWidth)
    },
    [setWidth]
  )

  const left = useMemo(
    () =>
      Math.max(
        Math.floor(
          Math.min(
            width * position - componentWidth / 2,
            width - componentWidth - 10
          )
        ),
        10
      ),
    [componentWidth, position, width]
  )
  return (
    <View
      style={{
        ...style,
        left,
        opacity: componentWidth ? 1 : 0,
        position: "absolute",
      }}
      onLayout={onLayout}
    >
      <Text style={{ color: props.color, fontSize: 12 }} weight="bold">
        {props.children}
      </Text>
    </View>
  )
}

const Labels = ({ width, isCard, color }): JSX.Element | null => {
  const { greatestX, greatestY, smallestX, smallestY } = useChartData()

  const { selectedCurrency } = useContext(DataContext)

  const currencySymbol = getCurrencySymbol(selectedCurrency)

  if (!greatestX) {
    return null
  }
  const positionMin = trim(
    (smallestY.x - smallestX.x) / (greatestX.x - smallestX.x)
  )
  const positionMax = trim(
    (greatestY.x - smallestX.x) / (greatestX.x - smallestX.x)
  )

  const smallestYVal = Number(smallestY.y)
  const largestYval = Number(greatestY.y)

  return (
    <>
      {positionMin ? (
        <CenteredLabel
          color={color}
          fontSize={isCard ? "13pt" : undefined}
          position={positionMin}
          style={{
            bottom: isCard ? -12 : "25%",
            zIndex: 6,
          }}
          width={width}
        >
          {`${currencySymbol}${smallestYVal.toFixed(7)}`}
        </CenteredLabel>
      ) : null}
      {positionMax ? (
        <CenteredLabel
          color={color}
          fontSize={isCard ? "13pt" : undefined}
          position={positionMax}
          style={{
            top: -10,
            left: isCard ? 0 : 50,
            zIndex: 6,
          }}
          width={width}
        >
          {`${currencySymbol}${largestYval.toFixed(7)}`}
        </CenteredLabel>
      ) : null}
    </>
  )
}

export default React.memo(Labels)
