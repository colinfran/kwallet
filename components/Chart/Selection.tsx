import React, { useContext, useEffect } from "react"
import {
  StyleSheet,
  TouchableWithoutFeedback,
  View,
  Text,
  Dimensions,
} from "react-native"
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withSpring,
} from "react-native-reanimated"
import { DataContext } from "../../providers/DataProvider"

export const Selection = ({
  graphData,
  arrayButtonsData,
  onButtonPress,
  selectedVariantIndex,
}): JSX.Element => {
  const offset = useSharedValue(0)

  const HEIGHT = 40
  const WIDTH = (Dimensions.get("window").width - 40) / arrayButtonsData.length

  const animatedStyles = useAnimatedStyle(() => {
    return {
      transform: [{ translateX: offset.value * WIDTH }],
    }
  })
  const { appColor, textColor, backgroundColor } = useContext(DataContext)

  const styles = StyleSheet.create({
    container: {
      width: "100%",
    },
    box: {
      width: WIDTH,
      height: HEIGHT,
      backgroundColor: backgroundColor,
      borderColor: appColor,
      borderRadius: 15,
      borderWidth: 2,
      justifyContent: "center",
      alignItems: "center",
      position: "absolute",
      left: 0,
      right: 0,
      margin: "auto",
      zIndex: 0,
      shadowOffset: {
        width: 0,
        height: 5,
      },
      shadowOpacity: 0.35,
      shadowRadius: 3.5,
      elevation: 5,
      shadowColor: textColor,
    },
    buttonList: {
      flexDirection: "row",
      width: "100%",
    },
    button: {
      height: HEIGHT,
      width: WIDTH,
      justifyContent: "center",
      alignItems: "center",
    },
    label: {
      fontSize: 16,
      color: "#5A5A5A",
      // textAlign: "center",
      zIndex: 5,
    },
  })

  return (
    <View style={styles.container}>
      <Animated.View style={[styles.box, animatedStyles]} />
      <View style={styles.buttonList}>
        {arrayButtonsData.map((element, i) => {
          const isSelected = i === selectedVariantIndex
          return (
            <TouchableWithoutFeedback
              key={i}
              testID={`test-chart-btn-${i}`}
              onPress={() => {
                onButtonPress(
                  graphData[element.buttonTitle.toLowerCase()].prices,
                  i
                )
                offset.value = withSpring(i)
              }}
            >
              <View style={styles.button}>
                <Text
                  style={[styles.label, isSelected && { color: textColor }]}
                >
                  {element.buttonTitle}
                </Text>
              </View>
            </TouchableWithoutFeedback>
          )
        })}
      </View>
    </View>
  )
}
