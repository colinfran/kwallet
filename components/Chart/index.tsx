import React, { useContext, useEffect, useState } from "react"
import { Dimensions, View, StyleSheet } from "react-native"
import {
  ChartDot,
  ChartPath,
  ChartPathProvider,
  ChartXLabel,
} from "@colinfran/animated-charts"
import * as Haptics from "expo-haptics"
import { Selection } from "./Selection"
import { DataContext } from "../../providers/DataProvider"

import Labels from "./Labels"
import { Skeleton } from "native-base"
import opacity from "hex-color-opacity"
import CustomYLabel from "./CustomYLabel"
import { getCurrencySymbol } from "../../constants/currencies"

export const { width: SIZE } = Dimensions.get("window")

const defaultData = [
  {
    x: 1679246714543,
    y: 0.014540020460445197,
  },
  {
    x: 1679247047789,
    y: 0.014571506674731024,
  },
  {
    x: 1679247288167,
    y: 0.014452072671740143,
  },
]

const Chart = (): JSX.Element => {
  const {
    graphData,
    appColor,
    selectedGraphIndex,
    setSelectedGraphIndex,
    textColor,
    selectedCurrency,
  } = useContext(DataContext)

  const currencySymbol = getCurrencySymbol(selectedCurrency)

  const isLoaded = graphData

  // const [selectedVariantIndex, setSelectedVariantIndex] = useState(0)

  const [data, setData] = useState(graphData?.day?.prices || defaultData)

  const [arrayButtonsData, setArrayButtonsData] = useState([
    { buttonTitle: "Day" },
  ])

  useEffect(() => {
    if (graphData) {
      if (graphData?.day?.prices) {
        setData(graphData?.day?.prices)
      }
      const arrayButtons = [{ buttonTitle: "Day" }]
      if (graphData?.week?.prices.length > 0)
        arrayButtons.push({ buttonTitle: "Week" })
      if (graphData?.month?.prices.length > 0)
        arrayButtons.push({ buttonTitle: "Month" })
      if (graphData?.year?.prices.length > 0)
        arrayButtons.push({ buttonTitle: "Year" })
      if (graphData?.all?.prices.length > 0)
        arrayButtons.push({ buttonTitle: "All" })
      setArrayButtonsData(arrayButtons)
    }
  }, [graphData, selectedCurrency])

  const invokeHaptic = (): void => {
    Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light)
  }

  const onButtonPress = (element, index): void => {
    setSelectedGraphIndex(index)
    setData(element)
    invokeHaptic()
  }

  const getX = (value): string => {
    "worklet"
    if (value && !isNaN(value)) {
      // return `${runOnJS(getTimeDate)(new Number(value))}`
      return `${new Date(new Number(value)).toLocaleString([], {
        dateStyle: "short",
        timeStyle: "short",
      })}`
    }
    return `${value}`
  }

  const [isActive, setIsActive] = useState(false)

  useEffect(() => {
    return () => {
      setSelectedGraphIndex(0)
    }
  }, [])

  return (
    <Skeleton
      // h={360}
      isLoaded={isLoaded}
      w={Dimensions.get("window").width}
      // style={{justifyContent:"center", alignItems:'center'}}
    >
      <View style={styles.container}>
        <View style={styles.buttonContainer}>
          <View style={styles.selectionContainer}>
            <Selection
              arrayButtonsData={arrayButtonsData}
              graphData={graphData}
              selectedVariantIndex={selectedGraphIndex}
              onButtonPress={onButtonPress}
            />
          </View>
        </View>
        <View>
          <ChartPathProvider
            data={{
              points: data,
              smoothingStrategy: "bezier",
            }}
            nowData={{
              currentPrice: graphData?.currentPrice || "",
              currencySymbol,
            }}
          >
            {!isActive && (
              <Labels color={textColor} isCard={false} width={SIZE} />
            )}
            <ChartPath
              backgroundGrandient={opacity(appColor, 0.2)}
              gradientEnabled={true}
              hapticsEnabled={true}
              height={Dimensions.get("window").height / 3}
              selectedStrokeWidth={3}
              stroke={appColor}
              strokeWidth={3}
              width={SIZE}
              onActiveFunction={() => setIsActive(true)}
              onEndFunction={() => setIsActive(false)}
            />
            <ChartDot size={16} style={{ backgroundColor: appColor }} />
            <View
              style={{
                // backgroundColor: opacity(appColor, 0.2),
                marginTop: -10,
              }}
            >
              <View
                style={{
                  justifyContent: "center",
                  alignItems: "center",
                  padding: 30,
                  marginTop: 20,
                }}
              >
                <ChartXLabel color={textColor} format={getX} />
                <CustomYLabel
                  color={textColor}
                  currencySymbol={currencySymbol}
                  currentPrice={graphData?.currentPrice || 0}
                  style={{ fontSize: 24 }}
                />
              </View>
            </View>
          </ChartPathProvider>
        </View>
      </View>
    </Skeleton>
  )
}

const styles = StyleSheet.create({
  container: {
    width: "100%",
    alignItems: "center",
    justifyContent: "center",
    gap: 40,
    marginTop: 10,
  },
  buttonContainer: {
    flexDirection: "row",
    justifyContent: "space-evenly",
    // padding: 20,
    paddingTop: 0,
    // marginBottom: 20,
  },
  selectionContainer: {
    justifyContent: "space-evenly",
  },
})

export default Chart
