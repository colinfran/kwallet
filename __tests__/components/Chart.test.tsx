import React from "react"
import { Providers, setSelectedMock } from "../helper"
import { fireEvent, render, screen } from "@testing-library/react-native"
import Chart from "../../components/Chart"
import { cleanup } from "@testing-library/react-native"

describe("<Chart />", () => {
  const RealDate = Date

  beforeEach(() => {
    // @ts-ignore Type 'typeof Date' is not assignable to type 'DateConstructor'
    global.Date = class extends RealDate {
      constructor() {
        super()
        return new RealDate("04/20/2023")
      }
    }
  })
  afterEach(() => {
    global.Date = RealDate
  })
  afterEach(cleanup)

  afterEach(() => {
    jest.restoreAllMocks()
  })

  it("renders correctly", () => {
    const tree = render(<Chart />, { wrapper: Providers }).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("buttons", () => {
    render(<Chart />, { wrapper: Providers })
    for (let i = 0; i < 5; i++) {
      const btn = screen.getByTestId(`test-chart-btn-${i}`)
      fireEvent.press(btn)
      expect(setSelectedMock.mock.lastCall[0]).toBe(i)
    }
    expect(setSelectedMock.mock.calls.length).toBe(5)
  })
})
