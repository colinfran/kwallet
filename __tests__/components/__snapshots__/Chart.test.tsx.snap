// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "gap": 40,
        "justifyContent": "center",
        "marginTop": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "paddingTop": 0,
        }
      }
    >
      <View
        style={
          {
            "justifyContent": "space-evenly",
          }
        }
      >
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <View
            animatedStyle={
              {
                "value": {
                  "transform": [
                    {
                      "translateX": 0,
                    },
                  ],
                },
              }
            }
            collapsable={false}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#654321",
                "borderColor": "#123456",
                "borderRadius": 15,
                "borderWidth": 2,
                "elevation": 5,
                "height": 40,
                "justifyContent": "center",
                "left": 0,
                "margin": "auto",
                "position": "absolute",
                "right": 0,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 5,
                  "width": 0,
                },
                "shadowOpacity": 0.35,
                "shadowRadius": 3.5,
                "transform": [
                  {
                    "translateX": 0,
                  },
                ],
                "width": 142,
                "zIndex": 0,
              }
            }
          />
          <View
            style={
              {
                "flexDirection": "row",
                "width": "100%",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 142,
                }
              }
              testID="test-chart-btn-0"
            >
              <Text
                style={
                  [
                    {
                      "color": "#5A5A5A",
                      "fontSize": 16,
                      "zIndex": 5,
                    },
                    {
                      "color": "#000",
                    },
                  ]
                }
              >
                Day
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 142,
                }
              }
              testID="test-chart-btn-1"
            >
              <Text
                style={
                  [
                    {
                      "color": "#5A5A5A",
                      "fontSize": 16,
                      "zIndex": 5,
                    },
                    false,
                  ]
                }
              >
                Week
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 142,
                }
              }
              testID="test-chart-btn-2"
            >
              <Text
                style={
                  [
                    {
                      "color": "#5A5A5A",
                      "fontSize": 16,
                      "zIndex": 5,
                    },
                    false,
                  ]
                }
              >
                Month
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 142,
                }
              }
              testID="test-chart-btn-3"
            >
              <Text
                style={
                  [
                    {
                      "color": "#5A5A5A",
                      "fontSize": 16,
                      "zIndex": 5,
                    },
                    false,
                  ]
                }
              >
                Year
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 142,
                }
              }
              testID="test-chart-btn-4"
            >
              <Text
                style={
                  [
                    {
                      "color": "#5A5A5A",
                      "fontSize": 16,
                      "zIndex": 5,
                    },
                    false,
                  ]
                }
              >
                All
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View>
      <View
        onLayout={[Function]}
        style={
          {
            "bottom": "25%",
            "left": 712,
            "opacity": 0,
            "position": "absolute",
            "zIndex": 6,
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontSize": 12,
            }
          }
          weight="bold"
        >
          $0.0144521
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={
          {
            "left": 435,
            "opacity": 0,
            "position": "absolute",
            "top": -10,
            "zIndex": 6,
          }
        }
      >
        <Text
          style={
            {
              "color": "#000",
              "fontSize": 12,
            }
          }
          weight="bold"
        >
          $0.0145715
        </Text>
      </View>
      <View
        animatedStyle={
          {
            "value": {},
          }
        }
        collapsable={false}
        forwardedRef={[Function]}
        handlerTag={1}
        handlerType="LongPressGestureHandler"
        onGestureHandlerEvent={[Function]}
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight="464.6666666666667"
          bbWidth="750"
          focusable={false}
          height={464.6666666666667}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 464,
                "width": 750,
              },
            ]
          }
          vbHeight={444.6666666666667}
          vbWidth={750}
          width={750}
        >
          <RNSVGGroup>
            <RNSVGPath
              d=""
              opacity={1}
              propList={
                [
                  "stroke",
                  "strokeWidth",
                ]
              }
              stroke={
                {
                  "payload": 4279383126,
                  "type": 0,
                }
              }
              strokeWidth="3"
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
      <View
        style={
          {
            "marginTop": -10,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "justifyContent": "center",
              "marginTop": 20,
              "padding": 30,
            }
          }
        >
          <TextInput
            animatedStyle={
              {
                "value": {},
              }
            }
            collapsable={false}
            color="#000"
            defaultValue="4/20/23, 12:00 AM"
            editable={false}
            text="4/20/23, 12:00 AM"
          />
          <TextInput
            animatedStyle={
              {
                "value": {},
              }
            }
            collapsable={false}
            color="#000"
            defaultValue="$0.0000000"
            editable={false}
            style={
              {
                "fontSize": 24,
              }
            }
            text="$0.0000000"
          />
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
